File: frontend/lib/model/order_model.dart
// Import the OrderDetail model

class Order {
  final String? id; // ma_don_hang
  final String customerId; // ma_khach_hang
  final DateTime orderDate; // ngay_dat
  final double totalAmount; // tong_tien - not null
  final String status; // trang_thai - default "chờ xử lý"
  final String customerName; // For display purposes
  final List<OrderDetail>? orderDetails; // Added order details list

  Order({
    this.id,
    required this.customerId,
    required this.orderDate,
    required this.totalAmount,
    this.status = "chờ xử lý", // Default value
    required this.customerName,
    this.orderDetails, // Added parameter for order details
  });

  // Create a copy of this Order with modified fields
  Order copyWith({
    String? id,
    String? customerId,
    DateTime? orderDate,
    double? totalAmount,
    String? status,
    String? customerName,
    List<OrderDetail>? orderDetails,
  }) {
    return Order(
      id: id ?? this.id,
      customerId: customerId ?? this.customerId,
      orderDate: orderDate ?? this.orderDate,
      totalAmount: totalAmount ?? this.totalAmount,
      status: status ?? this.status,
      customerName: customerName ?? this.customerName,
      orderDetails: orderDetails ?? this.orderDetails,
    );
  }

  // Create an Order from a Map
  factory Order.fromJson(Map<String, dynamic> json) {
    List<OrderDetail>? details;
    if (json['orderDetails'] != null) {
      details = List<OrderDetail>.from(
        (json['orderDetails'] as List).map(
          (detail) => OrderDetail.fromJson(detail),
        ),
      );
    }

    return Order(
      id: json['id'] as String?,
      customerId: json['customerId'] as String,
      orderDate: DateTime.parse(json['orderDate'] as String),
      totalAmount: (json['totalAmount'] as num).toDouble(),
      status: json['status'] as String? ?? "chờ xử lý",
      customerName: json['customerName'] as String,
      orderDetails: details,
    );
  }

  // Convert Order to a Map
  Map<String, dynamic> toJson() {
    final Map<String, dynamic> data = {
      'id': id,
      'customerId': customerId,
      'orderDate': orderDate.toIso8601String(),
      'totalAmount': totalAmount,
      'status': status,
      'customerName': customerName,
    };

    if (orderDetails != null) {
      data['orderDetails'] =
          orderDetails!.map((detail) => detail.toJson()).toList();
    }

    return data;
  }
}

// Define the OrderDetail class inline if needed
class OrderDetail {
  final String id; // ma_chi_tiet
  final String orderId; // ma_don_hang
  final String bookId; // ma_sach
  final int quantity; // so_luong
  final double price; // gia
  final String? bookTitle; // Optional book title for display

  OrderDetail({
    required this.id,
    required this.orderId,
    required this.bookId,
    required this.quantity,
    required this.price,
    this.bookTitle,
  });

  factory OrderDetail.fromJson(Map<String, dynamic> json) {
    return OrderDetail(
      id: json['id'].toString(),
      orderId: json['orderId'].toString(),
      bookId: json['bookId'].toString(),
      quantity: json['quantity'] as int,
      price: (json['price'] as num).toDouble(),
      bookTitle: json['bookTitle'] as String?,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'orderId': orderId,
      'bookId': bookId,
      'quantity': quantity,
      'price': price,
      'bookTitle': bookTitle,
    };
  }
}

File: backend_express/models/orderModel.js
const db = require('../config/database');

// Get all orders
exports.getAllOrders = async () => {
  try {
    const [rows] = await db.query(`
      SELECT o.ma_don_hang as id, o.ma_khach_hang as customerId, 
             o.ngay_dat as orderDate, o.tong_tien as totalAmount, 
             o.trang_thai as status, c.ten_khach_hang as customerName
      FROM orders o
      LEFT JOIN users c ON o.ma_khach_hang = c.ma_khach_hang
      ORDER BY o.ngay_dat DESC
    `);
    return rows;
  } catch (error) {
    console.error('Error getting all orders:', error);
    throw error;
  }
};

// Get a specific order by ID
exports.getOrderById = async (id) => {
  try {
    const [rows] = await db.query(`
      SELECT o.ma_don_hang as id, o.ma_khach_hang as customerId, 
             o.ngay_dat as orderDate, o.tong_tien as totalAmount, 
             o.trang_thai as status, c.ten_khach_hang as customerName
      FROM orders o
      LEFT JOIN users c ON o.ma_khach_hang = c.ma_khach_hang
      WHERE o.ma_don_hang = ?
    `, [id]);
    
    return rows.length ? rows[0] : null;
  } catch (error) {
    console.error('Error getting order by ID:', error);
    throw error;
  }
};

// Get a specific order by ID with details
exports.getOrderWithDetails = async (id) => {
  try {
    // First fetch the order
    const [orderRows] = await db.query(`
      SELECT o.ma_don_hang as id, o.ma_khach_hang as customerId, 
             o.ngay_dat as orderDate, o.tong_tien as totalAmount, 
             o.trang_thai as status, c.ten_khach_hang as customerName
      FROM orders o
      LEFT JOIN users c ON o.ma_khach_hang = c.ma_khach_hang
      WHERE o.ma_don_hang = ?
    `, [id]);
    
    if (orderRows.length === 0) {
      return null;
    }
    
    const order = orderRows[0];
    
    // Then fetch the order details
    const [detailRows] = await db.query(`
      SELECT d.ma_chi_tiet as id, d.ma_don_hang as orderId, 
             d.ma_sach as bookId, d.so_luong as quantity, 
             d.gia as price, b.ten_sach as bookTitle
      FROM oder_details d
      LEFT JOIN books b ON d.ma_sach = b.ma_sach
      WHERE d.ma_don_hang = ?
    `, [id]);
    
    // Add the details to the order
    order.orderDetails = detailRows;
    
    return order;
  } catch (error) {
    console.error('Error getting order with details:', error);
    throw error;
  }
};

// Create a new order
exports.createOrder = async (orderData) => {
  try {
    const { customerId, totalAmount, status = "chờ xử lý" } = orderData;
    orderData.orderDate = new Date();
    const [result] = await db.query(
      'INSERT INTO orders (ma_khach_hang, ngay_dat, tong_tien, trang_thai) VALUES (?, ?, ?, ?)',
      [customerId, orderData.orderDate, totalAmount, status]
    );
    
    return result.insertId;
  } catch (error) {
    console.error('Error creating order:', error);
    throw error;
  }
};

// Update an existing order
exports.updateOrder = async (id, orderData) => {
  try {
    const { customerId, orderDate, totalAmount, status } = orderData;
    
    const [result] = await db.query(
      'UPDATE orders SET ma_khach_hang = ?, ngay_dat = ?, tong_tien = ?, trang_thai = ? WHERE ma_don_hang = ?',
      [customerId, new Date(orderDate), totalAmount, status, id]
    );
    
    return result.affectedRows;
  } catch (error) {
    console.error('Error updating order:', error);
    throw error;
  }
};

// Update just the order status
exports.updateOrderStatus = async (id, status) => {
  try {
    const [result] = await db.query(
      'UPDATE orders SET trang_thai = ? WHERE ma_don_hang = ?',
      [status, id]
    );
    
    return result.affectedRows;
  } catch (error) {
    console.error('Error updating order status:', error);
    throw error;
  }
};

// Delete an order
exports.deleteOrder = async (id) => {
  try {
    // First delete related order details
    await db.query('DELETE FROM oder_details WHERE ma_don_hang = ?', [id]);
    
    // Then delete the order
    const [result] = await db.query('DELETE FROM orders WHERE ma_don_hang = ?', [id]);
    
    return result.affectedRows;
  } catch (error) {
    console.error('Error deleting order:', error);
    throw error;
  }
};

// Get order details for a specific order
exports.getOrderDetails = async (orderId) => {
  try {
    const [rows] = await db.query(`
      SELECT d.ma_chi_tiet as id, d.ma_don_hang as orderId, 
             d.ma_sach as bookId, d.so_luong as quantity, 
             d.gia as price, b.ten_sach as bookTitle
      FROM oder_details d
      LEFT JOIN books b ON d.ma_sach = b.ma_sach
      WHERE d.ma_don_hang = ?
    `, [orderId]);
    
    return rows;
  } catch (error) {
    console.error('Error getting order details:', error);
    throw error;
  }
};

// Add a detail to an order
exports.addOrderDetail = async (orderId, detailData) => {
  try {
    const { bookId, quantity, price } = detailData;
    
    const [result] = await db.query(
      'INSERT INTO oder_details (ma_don_hang, ma_sach, so_luong, gia) VALUES (?, ?, ?, ?)',
      [orderId, bookId, quantity, price]
    );
    
    // Update the order's total amount
    await db.query(
      `UPDATE orders 
       SET tong_tien = (SELECT SUM(so_luong * gia) FROM oder_details WHERE ma_don_hang = ?)
       WHERE ma_don_hang = ?`,
      [orderId, orderId]
    );
    
    return result.insertId;
  } catch (error) {
    console.error('Error adding order detail:', error);
    throw error;
  }
};
